Customizable note hit sounds & effects

FlooferLand opened the pull request:
This PR adds note hit sound effects, as well as a little sound system for playing general note-related sounds.
It also adds a settings toggle for note splash particles, as requested by a `// TODO` I saw.

See [FunkinCrew/funkin.assets/preload/sounds/noteSounds/how to.txt](https://github.com/FunkinCrew/funkin.assets/blob/a0e50669cfa0a0d10e16464b794b190b3eb55500/preload/sounds/noteSounds/how%20to.txt) to see how to add new sounds

### Video showcase
_GitHub doesn't allow videos above 10 MB so its behind a YouTube or Catbox link_

[![YouTube Thumbnail](http://i3.ytimg.com/vi/K-nh51LEa7U/hqdefault.jpg)](https://www.youtube.com/watch?v=K-nh51LEa7U)

[_(or a catbox.moe mirror)_](https://files.catbox.moe/c0sddm.webm)

_(NOTE: vine boom unfortunately had to be removed due to copyright concerns)_

---

**Associated assets PR:** https://github.com/FunkinCrew/funkin.assets/pull/36

_This PR was previously #2793._
_Made a new PR as I rewrote everything and the old PR contained remnants of the new settings widget system_

FlooferLand committed:
Add note splash settings toggle

Lasercar committed:
Chart Editor playtest option for Results Screen

FlooferLand committed:
Add note hit sounds and a general note audio playback system

Hundrec commented:
Please change the base branch to develop at the top of this PR!

amyspark-ng commented:
🔥🔥🔥🔥🔥

FlooferLand committed:
Merge branch 'develop' into note-effects

FlooferLand commented:
I always forget about the develop branch, my bad.
Not sure how to change my fork / local repo over to the `develop` branch without risking accidentally nuking the entire PR again, I hate Git.
Also, I solved a conflict and someone added a `showNotesplash` to `Strumline.hx`, but it doesn't seem to be used by anything?

Lasercar commented:
> I always forget about the `develop` branch, my bad.
> 
> Not sure how to change my fork / local repo over to the `develop` branch without risking accidentally nuking the entire PR again, I think I give the lead maintainers a heart attack any time they see me do anything with git lmao
> 
> Also, I solved a conflict and someone added a `showNotesplash` to `Strumline.hx`, but it doesn't seem to be used by anything?

Just start the branch using the develop one as a base, rather than main. Then, when you make the PR, make sure you're requesting to merge into the develop branch and you should be fine!

AbnormalPoof commented:
> I always forget about the `develop` branch, my bad.
> 
> Not sure how to change my fork / local repo over to the `develop` branch without risking accidentally nuking the entire PR again, I think I give the lead maintainers a heart attack any time they see me do anything with git lmao
> 
> Also, I solved a conflict and someone added a `showNotesplash` to `Strumline.hx`, but it doesn't seem to be used by anything?

`showNotesplash` is for mods. It's so that mods can disable/enable note splashes at will.

Lasercar committed:
Welcome back welcome music

AbnormalPoof commented:
This PR appears to have merge conflicts. Please fix them!

amyspark-ng committed:
fix notelay sound when not in grid

FlooferLand committed:
Merge branch 'develop' into note-effects

FlooferLand committed:
Merge branch 'develop' into note-effects

JackXson-Real reviewed (changes requested):
Works good! There are 2 changes that should be made, I will approve this once they are:
1) Remove vine boom sound effect. It's a joke hitsound that does not have any place in the base game + the sound is not public domain, so it would technically be illegal for them to put it in here without permission
2) Add a preference for toggling the strumline arrow highlight animation. (Basically make it function the same as how the opponent used to hit notes pre-0.3.0.
**This Animation**
![image](https://github.com/user-attachments/assets/9252d7c8-24c2-4535-b866-163a504f5db9)

FlooferLand committed:
Remove vine boom note sounds

It is sad to see a loved one go but it was sadly inevitable, may the vine boom rest in peace

FlooferLand committed:
Add note highlight settings toggle

FlooferLand commented:
Should be good now!
Sad to see the vine boom gone but I suspected it wouldn't make it in xD
More on the copyright of things if you're interested:
- The `Pool Ball` sound is an edited segment from a CC0 (public domain) sample I found on FreeSound.
- I have no idea where `Ping Pong` sound is but it's been circulating around and being included in several rhythm games without seemingly anyone knowing where it's originally from. I'm pretty sure its a legacy sound effect from the rhythm game Osu, but I doubt Peppy would issue a copyright strike for a decade old sound effect that is less than half a second long

Do note I made the highlight animation toggle turn off the highlight animations for both the player and the opponent's notes when its off; I could potentially make it so the player can chose whenever the highlight animation plays for them and not their opponent but ehh. This should be fine.

JackXson-Real commented:
> Should be good now! Sad to see the vine boom gone but I suspected it wouldn't make it in xD More on the copyright of things if you're interested:
> 
> * The `Pool Ball` sound is an edited segment from a CC0 (public domain) sample I found on FreeSound.
> * I have no idea where `Ping Pong` sound is but it's been circulating around and being included in several rhythm games without seemingly anyone knowing where it's originally from. I'm pretty sure its a legacy sound effect from the rhythm game Osu, but I doubt Peppy would issue a copyright strike for a decade old sound effect that is less than half a second long
> 
> Do note I made the highlight animation toggle turn off the highlight animations for both the player and the opponent's notes when its off; I could potentially make it so the player can chose whenever the highlight animation plays for them and not their opponent but ehh. This should be fine.

Yeah don't worry I meant to have it done on both player and opponent so it's fine. I will test this in a bit.

JackXson-Real reviewed (approved):
Thanks for removing the vine boom! Also the note highlight preference looks perfect, thanks for adding it! I hope this gets merged for 0.6.0

Lasercar committed:
A big commit for a small fix
Cursor in chart editor breaks in square after placing and removing note

Lasercar committed:
Custom note kind field fix
Now gets hidden if the note selected isn't a custom one

Hundrec commented:
Please resolve the merge conflicts now that the source code has been released!

FlooferLand committed:
Merge branch 'develop' into note-effects

FlooferLand committed:
Merge branch 'develop' into note-effects

Lasercar committed:
Chart - Fix notekinds changing on multi select

Also reverts a change by this author that caused selected events to be reset when the event toolbox is first opened.

FuroYT committed:
Change BGScrollingText to a single FlxText

BGScrollingText is now a single FlxText drawn multiple time in a single frame which is more optimized that a FlxSpriteGroup with multiple FlxSprites taking the text's graphic

FuroYT committed:
Added `funnyColor` backwards compatibility

FlooferLand committed:
Merge branch 'develop' into note-effects

EliteMasterEric committed:
Adjust WinAPI to allow setting theme on specific windows.

Co-authored-by: MAJigsaw77 <MAJigsaw77@users.noreply.github.com>

anysad committed:
Remove unused Chart Editor Context Menu code

KoloInDaCrib committed:
make volume settings optionally carry over to playtest

NotHyper-474 committed:
Fix a syntax error with the RegistryMacro and a partial fix for code completion

NotHyper-474 committed:
Add hardware keyboard detection on Android

NotHyper-474 committed:
fix: Fix incorrect NaN checks & simplify finite check

PurSnake committed:
Properly load variation scripts in the Chart Editor when playtesting

ninjamuffin99 committed:
more controls.hx cleanup, remove a few unused functions, and add `Void` return types to make checkstyle happier

ninjamuffin99 committed:
remove more unused vars in controls.hx

ninjamuffin99 committed:
remove keyboard "scheme" related code, completely unused

ninjamuffin99 committed:
remove the unused _P and _R stuff from the Action enum abstract

ninjamuffin99 committed:
update our custom FunkinAction check() function with newer logic from FlxAction

VirtuGuy committed:
Fix results screen scrolling text speed

AbnormalPoof committed:
Fix the nametag erroneously showing as "Boyfriend" when entering character select as Pico

AbnormalPoof committed:
Get the first owned character ID and use that for results

charlesisfeline committed:
null checkin for set health icon

VirtuGuy committed:
Fix anim editor onion skin offset problems

KarimAkra committed:
Fix stickers transition looking off-size when resizing the game during a song

EliteMasterEric committed:
Code cleanup and additional docs for WaveformSprite.

EliteMasterEric committed:
Implement a checkbox to hide the waveform, which improves FPS.

EliteMasterEric committed:
Pause the chart editor when focus is lost.

AbnormalPoof committed:
Re-export Pico's confirm backing card with BTA to bake the filters

Hundrec committed:
Don't play Stress Pico cutscene if charting

EliteMasterEric committed:
Add difficulty switches to the undo/redo history.

EliteMasterEric committed:
Switching difficulty now forces the selection to be cleared.

AbnormalPoof committed:
Update assets and add `getAtlasSettings()` for texture atlas characters

Co-authored-by: VirtuGuy <Wyatt2010shaw@gmail.com>

EliteMasterEric committed:
Fix a bug where cycling difficulties wasn't calculated properly when switching variations.

Hundrec committed:
Adjust lyrics for Monster and Winter Horrorland

Fix some errors and formatting

EliteMasterEric committed:
Implemented Clone Difficulty button in the Difficulty toolbox.

Fixes to Clone Difficulty dialog, and added Move Difficulty dialog.

assets submod

KoloInDaCrib committed:
fix starting a playtest opening pause menu

VirtuGuy committed:
Fix ghost tapping behavior when unpausing

moondroidcoder committed:
fix: Make lerping consistent across different FPS

AbnormalPoof committed:
Make Monster's `singLEFT-hold` offset consistent with `singLEFT`

AbnormalPoof committed:
Re-export THE CHUDDER!!!!

ninjamuffin99 committed:
use FlxMath.wrap() to make this cursor wrapping cleaner in CharSelectSubState

ninjamuffin99 committed:
clean up spamDirection code, use FlxDirectionFlags

ninjamuffin99 committed:
move character select cursor sprite code into it's own class

moondroidcoder committed:
Update lime ref

KoloInDaCrib committed:
better backwards compatibility for `FlxAtlasSprite` -> `FunkinSprite`

ninjamuffin99 committed:
early return for Chart Editor handleCursor()

AbnormalPoof committed:
Fix the cigarette not appearing in the Pico doppelganger cutscene

AbnormalPoof committed:
Fix the cursor lerping from the top left in character select

AbnormalPoof committed:
Adjust Darnell's idle animation

EliteMasterEric committed:
Update HXCPP to support building with C++20

MAJigsaw77 committed:
Remove `WINAPI_GetProcessMemoryWorkingSetSize` dublicate.

AbnormalPoof committed:
Revert "remove keyboard "scheme" related code, completely unused"

This reverts commit 7c8b7eee578b4c5caabf194d645b1473f233dd46.

AbnormalPoof committed:
Re-export Girlfriend to texture atlas

AbnormalPoof committed:
Re-export Boyfriend (Car) to texture atlas

MAJigsaw77 committed:
Improve library version checks.

AbnormalPoof committed:
Lower him a little

MAJigsaw77 committed:
Remove the yellow notice.

MAJigsaw77 committed:
Bump `hxvlc` to `2.2.5`.

AbnormalPoof committed:
Re-export the Parents in the Eggnog Erect cutscene with BTA

AbnormalPoof committed:
Re-export Pico's PERFECT rank animation with BTA

AbnormalPoof committed:
Re-export GF's character select sprites to remove a stray line

NotHyper-474 committed:
fix: Cursor not properly fading out when exiting CharSelect with backspace

EliteMasterEric committed:
Some optimization to highlighted note checks.

AbnormalPoof committed:
Add `getFramesWithKeyword()`

AbnormalPoof committed:
Bump `flixel-animate` and add `useRenderTexture` to `FunkinSprite` as an option.

NotHyper-474 committed:
fix: solve crash when deleting save data

AbnormalPoof committed:
Re-export the Freeplay difficulty stars using BTA

EliteMasterEric committed:
Clean up the Lime target config dropdown by removing unused targets like AIR and Flash.

EliteMasterEric committed:
Rewrite measure ticks to not get redrawn EVERY FRAME GODAM

ninjamuffin99 committed:
reduce code copypaste a tiny smidge for chart editor waveforms

ninjamuffin99 committed:
early return in playVocals() to remove a level of nesting

ninjamuffin99 committed:
fix char select inputs for keyb and controllers

ninjamuffin99 committed:
remove unused pitch stuff in CharSelect

EliteMasterEric committed:
Redo the backing color of the measure ticks.

EliteMasterEric committed:
Tweak vertical offset on the measure numbers.

ninjamuffin99 committed:
use UIStateMachine in FreeplayState instead of checking/modifying `controls.active` everywhere
thus making it easier to fix the "unlocked inputs" issue (fix FUNK-2021)

ninjamuffin99 committed:
change fpScoreDisplay to private variable

ninjamuffin99 committed:
change fp -> fpScoreDisplay, as a more descriptive variable name

moondroidcoder committed:
fix: Stop the game from muting in silent mode when you unfocus then focus

Co-Authored-By: Karim Akra <144803230+KarimAkra@users.noreply.github.com>

ninjamuffin99 committed:
remove all the letterSort.inputEnabled stuff, only use the FreeplayState's uiStateMachine to check if inputs should be enabled

ninjamuffin99 committed:
change capsuleOptionsMenu -> instSelectMenu in FreeplayState

ninjamuffin99 committed:
remove some redundant getControls() in FreeplayState, just access controls directly

ninjamuffin99 committed:
remove some redundant getControls() in FreeplayState, just access controls directly

AbnormalPoof committed:
Add global offsets for Pico and Nene's pixel variants

Lasercar committed:
Chart Editor - additional audio checks for vocal only songs

sector-a committed:
Don't add the current difficulty in `Difficulty` entry of the Pause Menu.

Revert "Don't add the current difficulty in `Difficulty` entry of the Pause Menu."

This reverts commit e0518f636dceeafa6dd00c922972afa5840e1161.

Do nothing if you change difficulty to the same one.

moondroidcoder committed:
Decrease blocksize compression for gfChill

VirtuGuy committed:
Fix difficulty toolbox showing wrong song name

AbnormalPoof committed:
Shift the flames how did this happen

moondroidcoder committed:
Exclude gf Texture Atlases from ASTC Compression

AbnormalPoof committed:
Re-export Tankman and Tankman (Bloody) to texture atlas

AbnormalPoof committed:
Make Tankman less shiny

AbnormalPoof committed:
Re-export `bf` and `bf-dark` to texture atlas form and sever the death animation from `bf-car`

AbnormalPoof committed:
De-sharpen some atlases and optimize Monster's atlas

AbnormalPoof committed:
Re-export Spooky Kids to remove a billion black lines

AbnormalPoof committed:
Re-export Girlfriend (Dark) to texture atlas

AbnormalPoof committed:
Re-export some Week 7 assets to texture atlases

KoloInDaCrib committed:
add frames to the event data toolbox

AbnormalPoof committed:
Re-export Pico (holding Nene) to texture atlas

AbnormalPoof committed:
Re-export Otis (Speaker Shooter) to texture atlas

Kade-github committed:
Crash Fix and Visual Bug Fix on InstSelect

Hundrec committed:
Widen Events toolbox

KarimAkra committed:
Update lime ref to hide the ios home indicator

moondroidcoder committed:
Exclude pico holding nene from astc compression

AbnormalPoof committed:
Fine-tuned offsets for 2 characters

AbnormalPoof committed:
Re-export the background Tankmen in Week 7 Erect as texture atlases

moondroidcoder committed:
Decrease blocksizes for rando and sniper in Week 7 Erect

Hundrec committed:
Remove Chart Editor tick trace and bump assets

AbnormalPoof committed:
Fix atlases moving to the wrong place on restart

Bundled with a ton of manual offset fixes and adjustments

FlooferLand committed:
Merge branch 'develop' into note-effects

Hundrec commented:
Review requests don't matter too much, but feel free to request them
I have no idea when this will be reviewed, but we *might* ask you to reopen a new PR for this in case it can't be rebased cleanly

FlooferLand commented:
Fair. The PR was previously reviewed, and should be able to be rebased pretty cleanly on merge?
I've been fixing merge conflicts for about a year now though hoping it'd be in the game someday, and it's been pending triage for almost a year now which is pretty strange to me

EliteMasterEric commented:
After some internal discussion, we've concluded that if this feature is accepted, we will implement it ourselves internally and develop custom sounds for it, rather than accepting an outside contribution. Thank you for taking the time to develop and submit this regardless.
